StorageUtility/Linux/StorageUtility.cpp

```diff
@@ -16,252 +16,177 @@ limitations under the License.
+        String udeviceNode = udev_device_get_devnode( udevDevice );
+        sStorageAdapterProperty property;
+
+        if( true == InitStorageAdapterProperty( property, udevDevice ) )
         {
-            error = GetLastError();
-            if ( ERROR_INSUFFICIENT_BUFFER != error )
-            {
-                HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, pDevDetailData);
-                if ( eOnErrorBehavior::Continue == OnErrorBehavior ) { continue; }
-                SetupDiDestroyDeviceInfoList( devInfoHandle );
-                return( eErrorCode::Unknown );
-            }
+            g_MappingDevicePathAndDeviceHandle->MappingDevicePathToProperties[udeviceNode] = property;
+            DevicePaths.push_back( udeviceNode );
         }
 
-        eErrorCode errorCode;
-        errorCode = eErrorCode::Unknown;
-        Callback.Function(Callback.Data, devInfoHandle, devInfoData, devInterfaceData, pDevDetailData, bufferSize, errorCode);
-
-        HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, pDevDetailData);
+        udev_device_unref( udevDevice );
     }
+    /* Free the enumerator object */
+    udev_enumerate_unref( enumerate );
 
-    SetupDiDestroyDeviceInfoList( devInfoHandle );
-
-    return( eErrorCode::None );
-}*/
+    udev_unref( udevObject );
+}
 
-eErrorCode GetStorageAdapterProperty( HANDLE Handle, sStorageAdapterProperty& AdapterProperty )
+void GetStorageDeviceHandle( const String& DevicePath, DeviceHandle& Handle )
 {
-   /* STORAGE_DESCRIPTOR_HEADER storageDescHeader;
-
-    STORAGE_PROPERTY_QUERY StorageProperty;
-    StorageProperty.PropertyId = StorageAdapterProperty;
-    StorageProperty.QueryType = PropertyStandardQuery;
-
-    DWORD bytesReturned;
-    DWORD ret =
-        DeviceIoControl(
-                        Handle,
-                        IOCTL_STORAGE_QUERY_PROPERTY,
-                        &StorageProperty,
-                        sizeof(STORAGE_PROPERTY_QUERY),
-                        &storageDescHeader,
-                        sizeof(STORAGE_DESCRIPTOR_HEADER),
-                        &bytesReturned,
-                        NULL
-                        );
-    if ( 0 == ret )
+    Handle = open( DevicePath.c_str(), O_RDONLY|O_NONBLOCK );
+    if ( 0 > Handle )
     {
-        return( eErrorCode::Io );
+        throw std::runtime_error( "ERROR: Cannot open device \n" );
     }
+    g_MappingDevicePathAndDeviceHandle->MappingDeviceHandleToSysPath[Handle] = DevicePath;
+}
 
-    if ( 0 == storageDescHeader.Size )
+eErrorCode GetStorageAdapterProperty( DeviceHandle Handle, sStorageAdapterProperty& AdapterProperty )
+{
+    if (  g_MappingDevicePathAndDeviceHandle->MappingDeviceHandleToSysPath.find( Handle ) ==  g_MappingDevicePathAndDeviceHandle->MappingDeviceHandleToSysPath.end() )
     {
-        return( eErrorCode::Io );   //TODO: replace with more descriptive error
+        throw std::runtime_error( "ERROR: Cannot found mapping of DevicePath and Handle \n" );
```

Comment:
I think it is finding of maps. If not found a map of Device Path and Device Handle then throw a exception not return error. I read code on Window again and saw the same (eErrorCode::Io). I will use return eErrorCode::Io. Thanks a.Bac